// workorbit-central-logger/src/config/loggerConfig.js
const { format, transports } = require("winston");
const DailyRotateFile = require("winston-daily-rotate-file");
const path = require("path");

// ✅ Console Logging Format (For Development)
const consoleFormat = format.combine(
    format.colorize(),
    format.timestamp({ format: "YYYY-MM-DD HH:mm:ss" }),
    format.printf(({ timestamp, level, message }) => `${timestamp} [${level.toUpperCase()}]: ${message}`)
);

// ✅ JSON Log Format (For Production & File Logs)
const jsonLogFormat = format.combine(
    format.timestamp({ format: "YYYY-MM-DD HH:mm:ss" }),
    format.json()
);

// ✅ Daily Rotating File Transport (Logs all levels)
const fileRotateTransport = new DailyRotateFile({
    filename: path.join(__dirname, "../logs/application-%DATE%.log"),
    datePattern: "YYYY-MM-DD",
    zippedArchive: true, // Compress old log files
    maxSize: "20m", // Maximum size per log file
    maxFiles: "30d", // Retain logs for 30 days
    format: jsonLogFormat // Use JSON format for structured logs
});

// ✅ Separate Error Logs (For Better Error Monitoring)
const errorFileTransport = new DailyRotateFile({
    filename: path.join(__dirname, "../logs/error-%DATE%.log"),
    level: "error",
    datePattern: "YYYY-MM-DD",
    zippedArchive: true,
    maxSize: "20m",
    maxFiles: "30d",
    format: jsonLogFormat
});

// ✅ Export Logging Configuration
module.exports = {
    fileRotateTransport,
    errorFileTransport,
    consoleFormat,
    jsonLogFormat
};
