// workorbit-central-logger/src/services/LoggerService.js
const winston = require("winston");
const path = require("path");
const fs = require("fs");
const axios = require("axios");
const { v4: uuidv4 } = require("uuid");
require("winston-daily-rotate-file");

// ✅ Read environment variables
const ENABLE_SLACK_ALERTS = process.env.SLACK_WEBHOOK_URL ? true : false;
const ENABLE_JSON_LOGGING = process.env.JSON_LOGGING === "true";
const ENABLE_DEBUG_MODE = process.env.DEBUG === "true";
const DISABLE_CONSOLE_LOGS = process.env.DISABLE_CONSOLE_LOGS === "true"; // 🔥 Disable console logs in production

class LoggerService {
    constructor(serviceName = "Application") {
        this.serviceName = serviceName;
        this.logDir = path.join(process.cwd(), "logs");

        // ✅ Ensure log directory exists
        if (!fs.existsSync(this.logDir)) {
            fs.mkdirSync(this.logDir, { recursive: true });
        }

        this.logLevel = process.env.LOG_LEVEL || "info";

        // ✅ Define log format (Supports both text & JSON)
        const logFormat = ENABLE_JSON_LOGGING
            ? winston.format.json()
            : winston.format.printf(({ timestamp, level, message, service, requestId }) => {
                  return `${timestamp} [${service}]${requestId ? ` [RequestID: ${requestId}]` : ""} ${level.toUpperCase()}: ${message}`;
              });

        // ✅ Lazy initialization of transports
        this.transports = this._createTransports();

        this.logger = winston.createLogger({
            level: this.logLevel,
            defaultMeta: { service: this.serviceName },
            format: winston.format.combine(
                winston.format.timestamp({ format: "YYYY-MM-DD HH:mm:ss" }),
                logFormat
            ),
            transports: this.transports,
        });

        if (!DISABLE_CONSOLE_LOGS) {
            console.log(`🛠️ Logger initialized at level: ${this.logLevel}`);
        }
    }

    /**
     * ✅ Creates transport instances for logging.
     */
    _createTransports() {
        const transports = [];

        // ✅ Console logging (only enabled in non-production or when DEBUG is on)
        if (!DISABLE_CONSOLE_LOGS) {
            transports.push(
                new winston.transports.Console({
                    format: winston.format.combine(
                        winston.format.colorize(),
                        winston.format.simple()
                    ),
                })
            );
        }

        // ✅ Daily log rotation
        transports.push(
            new winston.transports.DailyRotateFile({
                filename: path.join(this.logDir, "app-%DATE%.log"),
                datePattern: "YYYY-MM-DD",
                zippedArchive: true,
                maxSize: "20m",
                maxFiles: "30d",
            })
        );

        // ✅ Separate file for error logs
        transports.push(
            new winston.transports.File({
                filename: path.join(this.logDir, "errors.log"),
                level: "error",
                maxsize: 5 * 1024 * 1024,
                maxFiles: 5,
                tailable: true,
            })
        );

        return transports;
    }

    log(level, message, requestId = null) {
        this.logger.log({ level, message, requestId });
        if ((level === "error" || level === "warn") && ENABLE_SLACK_ALERTS) {
            this._sendWebhookAlert(level, message, requestId);
        }
    }

    info(message, requestId = null) {
        this.logger.info({ message, requestId });
    }

    warn(message, requestId = null) {
        this.logger.warn({ message, requestId });
        if (ENABLE_SLACK_ALERTS) this._sendWebhookAlert("warn", message, requestId);
    }

    error(message, requestId = null) {
        this.logger.error({ message, requestId });
        if (ENABLE_SLACK_ALERTS) this._sendWebhookAlert("error", message, requestId);
    }

    debug(message, requestId = null) {
        if (ENABLE_DEBUG_MODE) {
            this.logger.debug({ message, requestId });
        }
    }

    /**
     * ✅ Sends an alert to Slack or Webhook when an error occurs.
     */
    async _sendWebhookAlert(level, message, requestId = null) {
        if (!ENABLE_SLACK_ALERTS) return;

        const payload = {
            text: `🚨 *${level.toUpperCase()} ALERT*\n*Service:* ${this.serviceName}\n*Request ID:* ${requestId || "N/A"}\n*Message:* ${message}`,
        };

        try {
            await axios.post(process.env.SLACK_WEBHOOK_URL, payload);
            console.log(`✅ Slack alert sent for ${level.toUpperCase()} log.`);
        } catch (error) {
            console.error("❌ Failed to send Slack alert:", error.message);
        }
    }
}

// ✅ Export both the class and a default instance
module.exports = {
    LoggerService,
    logger: new LoggerService(),
    generateRequestId: uuidv4, // ✅ Utility function for generating Request IDs
};
